File: metasystem/scripts/agents/open-work.py
#!/usr/bin/env python3
"""Report plans that name an unblocked next step while nothing is running.

Continuation is the one part of the loop no prompt can guarantee. An agent that
ends its turn with work still named in a plan has not been disobedient, it has
simply stopped, and prose telling it not to is not a mechanism. This reporter is
the mechanism: the end-of-turn hook runs it, so stopping with open work becomes
a visible fact rather than something only a reader would notice.

Its inputs are the structured fields `plans/README.md` already mandates for
every stream, not free prose: the `- Next step:` line, the
`- Waiting on the human:` line, and the `- In flight right now:` line. A plan is
open work when it names a next step, is not waiting on the human, and no agent
job is in flight.

It also reports a plan whose own account of itself contradicts the job records,
because a check that reads a stale plan reports the wrong work as confidently as
the right work, which is worse than staying quiet.

Nothing here is specific to a runtime. It reads files written by the dispatcher
and by humans, so it produces the same answer under Claude, Codex, Devin, or no
agent at all. How the answer reaches an agent differs by runtime; what the
answer is does not.
"""

from __future__ import annotations

import argparse
import json
import subprocess
import os
import sys
import time
import re
from pathlib import Path

# A next step that says nothing is left. Anything else is real work.
SETTLED = re.compile(
    r"^(none|nothing|n/?a|done|complete[d]?|-|tbd)\b[.\s]*$", re.IGNORECASE
)
# A waiting line that names no blocker. Anything else blocks continuation.
UNBLOCKED = re.compile(r"^(none|nothing)\b", re.IGNORECASE)
IN_FLIGHT = {"pending", "running"}


def field(text: str, label: str) -> str | None:
    """Return the value of a mandated `- <label>:` line, if present."""
    match = re.search(rf"^-\s*{re.escape(label)}\s*:\s*(.*)$", text, re.MULTILINE)
    if match is None:
        return None
    return match.group(1).strip()


def jobs_in_flight(harness_root: Path) -> int:
    count = 0
    for path in (harness_root / "artifacts" / "agents" / "jobs").glob("*.json"):
        try:
            record = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            # An unreadable record is the census's problem, not this reporter's.
            continue
        if record.get("status") in IN_FLIGHT:
            count += 1
    if count == 0 and gates_running(harness_root):
        count += 1
    return count


def gates_running(harness_root: Path) -> bool:
    """The orchestrator's own gate runs are work in flight that no job record
    describes: they run as background commands of the host session. A plan
    that says "waiting for the gates" while the gates are running is accurate,
    and calling it stale teaches people to stop writing the truth into plans.

    A gate says so itself, in a marker naming its process. Asking `pgrep -f`
    instead matched anything that merely MENTIONED the gate script -- a
    wait-loop, a grep, this session's own shell -- and answered for the whole
    machine rather than for this checkout. It reported "STILL WORKING: the test
    gates" with no gate running anywhere, and, because a running gate counts as
    work in flight, a false yes silences the open-work report entirely.
    """
    # Fixtures state the answer rather than racing whatever else is on the
    # machine; the suite that runs them is itself a gate run.
    declared = os.environ.get("METASYSTEM_GATES_RUNNING")
    if declared in {"0", "1"}:
        return declared == "1"
    helper = Path(__file__).resolve().parent / "gate-run.py"
    if not helper.is_file():
        return False
    probe = subprocess.run(
        [sys.executable, str(helper), "check", "--root", str(harness_root)],
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
        text=True,
        check=False,
    )
    return probe.returncode == 0 and probe.stdout.strip() == "1"


def stale_plans(harness_root: Path) -> list[str]:
    """Report plans whose own state contradicts the job records.

    A plan that misdescribes reality makes every check reading it worse than
    useless: it reports the wrong work with the same confidence as the right
    work. This session lost a turn to exactly that, so the accuracy of the
    signal is checked alongside the signal.
    """
    live = set()
    chains: dict[str, dict] = {}
    now = time.time()
    grace = float(os.environ.get("METASYSTEM_CHAIN_GRACE_SECONDS", "5400"))
    for path in (harness_root / "artifacts" / "agents" / "jobs").glob("*.json"):
        try:
            record = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            continue
        job_id = record.get("jobId", "") or path.stem
        if record.get("status") in IN_FLIGHT:
            live.add(job_id)
        root = re.sub(r"-r[0-9]+$", "", job_id)
        entry = chains.setdefault(root, {"ids": set(), "closed": False, "newest": None})
        entry["ids"].add(job_id)
        if job_id == root and record.get("chainClosed") is True:
            entry["closed"] = True
        match = re.search(r"-r([0-9]+)$", job_id)
        rank = int(match.group(1)) if match else 1
        newest = entry["newest"]
        if newest is None or rank > newest[0]:
            try:
                mtime = path.stat().st_mtime
            except OSError:
                mtime = 0.0
            entry["newest"] = (rank, record.get("status"), mtime)

    # IL-16: an open chain is work in flight for a PLAN's purposes even while no
    # round is running. Between rounds the orchestrator is adjudicating, and the
    # false stale-plan reports that gap produced taught the operator to ignore a
    # true one. The window is bounded so an abandoned chain still goes stale,
    # and jobs_in_flight() deliberately keeps the strict view: the stop hook
    # must still refuse a turn that walks away from an open chain.
    current = set(live) | {re.sub(r"-r[0-9]+$", "", job) for job in live}
    for root, entry in chains.items():
        if entry["closed"] or entry["newest"] is None:
            continue
        rank, status, mtime = entry["newest"]
        if status == "completed" and now - mtime <= grace:
            current.add(root)
            current |= entry["ids"]

    lines = []
    for plan in sorted((harness_root / "plans").glob("*.md")):
        if plan.name == "README.md":
            continue
        try:
            text = plan.read_text(encoding="utf-8")
        except OSError:
            continue
        claim = field(text, "In flight right now")
        if claim is None or not claim or UNBLOCKED.match(claim):
            # A stream saying it has nothing running is not contradicted by some
            # other stream's job. Only a claim of its own in-flight work can be
            # wrong, and flagging every idle plan whenever anything runs would
            # make this noise rather than signal.
            continue
        name = plan.relative_to(harness_root)
        if "gate" in claim.lower() and gates_running(harness_root):
            # The claim names the gates and the gates are running: accurate.
            continue
        if not current:
            lines.append(
                f"STALE-PLAN {name}: claims work in flight while no job is running"
            )
            continue
        if not any(job and job in claim for job in current):
            lines.append(
                f"STALE-PLAN {name}: names in-flight work that is not among the "
                f"running jobs or open chains"
            )
    return lines


def open_work(harness_root: Path) -> list[str]:
    if jobs_in_flight(harness_root):
        # Work is in flight; the turn is not idle and nothing is being dropped.
        return []
    lines = []
    for plan in sorted((harness_root / "plans").glob("*.md")):
        # README.md documents the stream format and carries the field names
        # inside a fenced example; it is not itself a stream.
        if plan.name == "README.md":
            continue
        try:
            text = plan.read_text(encoding="utf-8")
        except OSError:
            continue
        step = field(text, "Next step")
        if not step or SETTLED.match(step):
            continue
        waiting = field(text, "Waiting on the human")
        if waiting and not UNBLOCKED.match(waiting):
            continue
        name = plan.relative_to(harness_root)
        lines.append(f"OPEN-WORK {name}: {step}")
    return lines


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", required=True, help="metasystem root")
    args = parser.parse_args()
    root = Path(args.repo).resolve()
    if not (root / "plans").is_dir():
        return 0
    for line in stale_plans(root):
        print(line)
    for line in open_work(root):
        print(line)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: metasystem/internal/report/openwork.go
package report

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/widoriezebos/agentic-tools/metasystem/internal/atomicfile"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/brain"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/dispatch"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/gaterun"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/goal"
	"golang.org/x/sys/unix"
)

// OpenWork reports plans that name an unblocked next step while nothing is
// running, and plans whose own account of themselves contradicts the job
// records. Continuation is the one part of the loop no prompt can guarantee,
// so this reporter makes stopping with open work a visible fact. It reads only
// the structured fields plans/README.md mandates, so its answer is the same
// under any runtime or none.
func OpenWork(root string) []string {
	root = resolveRepo(root)
	if info, err := os.Stat(filepath.Join(root, "plans")); err != nil || !info.IsDir() {
		return nil
	}
	return append(stalePlans(root), openWork(root)...)
}

// now and grace are the time source and chain grace window, overridable in tests.
var now = time.Now

func graceSeconds() float64 {
	if v := os.Getenv("METASYSTEM_CHAIN_GRACE_SECONDS"); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			return f
		}
	}
	return 5400
}

var (
	settledStep    = regexp.MustCompile(`(?i)^(none|nothing|n/?a|done|completed?|-|tbd)\b[.\s]*$`)
	templateValue  = regexp.MustCompile(`^<[^<>\r\n]+>$`)
	unblockedField = regexp.MustCompile(`(?i)^(none|nothing)\b`)
	roundSuffix    = regexp.MustCompile(`-r[0-9]+$`)
	roundRank      = regexp.MustCompile(`-r([0-9]+)$`)
)

var (
	inFlightStatus      = map[string]bool{"pending": true, "running": true}
	brainInFlightStatus = map[string]bool{"pending-setup": true, "pending": true, "running": true}
)

func inFlightStatuses(root string) map[string]bool {
	state := brain.Read(root, goal.ExistingLedgerIdentity(root))
	if state.State == brain.Declared || state.State == brain.Corrupt {
		return brainInFlightStatus
	}
	return inFlightStatus
}

func resolveRepo(root string) string {
	if abs, err := filepath.Abs(root); err == nil {
		root = abs
	}
	if resolved, err := filepath.EvalSymlinks(root); err == nil {
		return resolved
	}
	return root
}

// planField returns the value of a mandated "- <label>:" line, if present.
func planField(text, label string) (string, bool) {
	re := regexp.MustCompile(`(?m)^-\s*` + regexp.QuoteMeta(label) + `\s*:\s*(.*)$`)
	m := re.FindStringSubmatch(text)
	if m == nil {
		return "", false
	}
	return strings.TrimSpace(m[1]), true
}

func readJobRecords(root string) []map[string]any {
	paths, _ := filepath.Glob(filepath.Join(root, "artifacts/agents/jobs", "*.json"))
	var records []map[string]any
	for _, path := range paths {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		var record map[string]any
		if json.Unmarshal(data, &record) != nil {
			continue
		}
		record["__path"] = path
		records = append(records, record)
	}
	return records
}

func jobsInFlight(root string) int {
	count := 0
	records := readJobRecords(root)
	statuses := inFlightStatuses(root)
	for _, record := range records {
		if status, _ := record["status"].(string); statuses[status] {
			count++
		}
	}
	if len(openChainsInFlight(records)) > 0 {
		count++
	}
	if count == 0 && gatesRunning(root) {
		count++
	}
	return count
}

func openChainsInFlight(records []map[string]any) []string {
	chains := map[string]*chainEntry{}
	for _, record := range records {
		path, _ := record["__path"].(string)
		jobID, _ := record["jobId"].(string)
		if jobID == "" {
			jobID = strings.TrimSuffix(filepath.Base(path), ".json")
		}
		rootJob := roundSuffix.ReplaceAllString(jobID, "")
		entry := chains[rootJob]
		if entry == nil {
			entry = &chainEntry{}
			chains[rootJob] = entry
		}
		if jobID == rootJob {
			if closed, ok := record["chainClosed"].(bool); ok {
				entry.rootOpen = !closed
			}
		}
		rank := 1
		if m := roundRank.FindStringSubmatch(jobID); m != nil {
			rank, _ = strconv.Atoi(m[1])
		}
		if !entry.hasNewest || rank > entry.newestRank {
			entry.hasNewest = true
			entry.newestRank = rank
			entry.status, _ = record["status"].(string)
		}
	}
	var roots []string
	for rootJob, entry := range chains {
		if entry.rootOpen && entry.hasNewest && !dispatch.TerminalStatus(entry.status) {
			roots = append(roots, rootJob)
		}
	}
	sort.Strings(roots)
	return roots
}

// gatesRunning reports whether a gate is in flight — from the fixture override
// when set, otherwise from the gate-run markers this checkout keeps.
func gatesRunning(root string) bool {
	switch os.Getenv("METASYSTEM_GATES_RUNNING") {
	case "1":
		return true
	case "0":
		return false
	}
	return gaterun.Running(root)
}

type chainEntry struct {
	ids        map[string]bool
	closed     bool
	rootOpen   bool
	hasNewest  bool
	newestRank int
	status     string
	mtime      time.Time
}

type openWorkSeenRecord struct {
	SchemaVersion int                                     `json:"schemaVersion"`
	Plans         map[string]map[string]openWorkSeenEntry `json:"plans"`
}

type openWorkSeenEntry struct {
	Line    string `json:"line"`
	FirstAt string `json:"firstAt"`
}

func openWorkSeenPath(root string) string {
	return filepath.Join(resolveRepo(root), "artifacts", "agents", "supervision", "open-work-seen.json")
}

func openWorkSeenWarning(path string, err error) string {
	return fmt.Sprintf("OPEN-WORK-SEEN-RESET %s: %v", path, err)
}

func readOpenWorkSeen(path string) (openWorkSeenRecord, string) {
	empty := openWorkSeenRecord{SchemaVersion: 1, Plans: map[string]map[string]openWorkSeenEntry{}}
	data, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return empty, ""
	}
	if err != nil {
		return empty, openWorkSeenWarning(path, err)
	}
	var record openWorkSeenRecord
	if err := json.Unmarshal(data, &record); err != nil {
		return empty, openWorkSeenWarning(path, err)
	}
	if record.SchemaVersion != 1 {
		return empty, openWorkSeenWarning(path, fmt.Errorf("unexpected schema version %d", record.SchemaVersion))
	}
	if record.Plans == nil {
		return empty, openWorkSeenWarning(path, fmt.Errorf("plans map is null"))
	}
	for plan, entries := range record.Plans {
		if entries == nil {
			return empty, openWorkSeenWarning(path, fmt.Errorf("plan %s has a null entries map", plan))
		}
		for digest, entry := range entries {
			expected := fmt.Sprintf("%x", sha256.Sum256([]byte(entry.Line)))
			if digest != expected {
				return empty, openWorkSeenWarning(path, fmt.Errorf("plan %s contains a line digest mismatch", plan))
			}
		}
	}
	return record, ""
}

// OpenWorkSeenWarning reads the durable marker without changing it. A
// deadline refusal uses this path so it cannot spend an open-work refusal.
func OpenWorkSeenWarning(root string) string {
	_, warning := readOpenWorkSeen(openWorkSeenPath(root))
	return warning
}

// MarkOpenWorkSeen durably records the current open plan lines, prunes lines
// no longer present, and tells the verdict which lines already refused a turn.
func MarkOpenWorkSeen(root string, items []goal.Item, at time.Time) ([]goal.Item, string, error) {
	path := openWorkSeenPath(root)
	if len(items) == 0 {
		if _, err := os.Stat(path); os.IsNotExist(err) {
			return items, "", nil
		}
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, "", fmt.Errorf("prepare open-work seen directory: %w", err)
	}
	lockFile, err := os.OpenFile(path+".lock", os.O_CREATE|os.O_RDWR, 0o600)
	if err != nil {
		return nil, "", fmt.Errorf("open open-work seen lock: %w", err)
	}
	defer lockFile.Close()
	if err := unix.Flock(int(lockFile.Fd()), unix.LOCK_EX); err != nil {
		return nil, "", fmt.Errorf("lock open-work seen record: %w", err)
	}
	defer func() { _ = unix.Flock(int(lockFile.Fd()), unix.LOCK_UN) }()

	record, warning := readOpenWorkSeen(path)

	marked := append([]goal.Item(nil), items...)
	current := openWorkSeenRecord{SchemaVersion: 1, Plans: map[string]map[string]openWorkSeenEntry{}}
	stamp := at.UTC().Format(time.RFC3339)
	for index := range marked {
		plan := marked[index].Id
		line := marked[index].FullDetail
		if line == "" {
			line = marked[index].Detail
		}
		digest := marked[index].LineDigest
		if digest == "" {
			digest = fmt.Sprintf("%x", sha256.Sum256([]byte(line)))
		}
		marked[index].LineDigest = digest
		entries := record.Plans[plan]
		entry, seen := entries[digest]
		marked[index].PreviouslyRefused = seen
		if !seen {
			entry = openWorkSeenEntry{Line: line, FirstAt: stamp}
		}
		current.Plans[plan] = map[string]openWorkSeenEntry{digest: entry}
	}
	encoded, err := json.MarshalIndent(current, "", "  ")
	if err != nil {
		return nil, "", fmt.Errorf("render open-work seen record: %w", err)
	}
	if _, err := atomicfile.WriteText(path, string(encoded)+"\n", ""); err != nil {
		return nil, "", fmt.Errorf("write open-work seen record: %w", err)
	}
	return marked, warning, nil
}

func stalePlans(root string) []string {
	return stalePlansWithStatuses(root, inFlightStatuses(root))
}

func stalePlansWithStatuses(root string, statuses map[string]bool) []string {
	live := map[string]bool{}
	chains := map[string]*chainEntry{}
	for _, record := range readJobRecords(root) {
		path, _ := record["__path"].(string)
		jobID, _ := record["jobId"].(string)
		if jobID == "" {
			jobID = strings.TrimSuffix(filepath.Base(path), ".json")
		}
		status, _ := record["status"].(string)
		if statuses[status] {
			live[jobID] = true
		}
		rootJob := roundSuffix.ReplaceAllString(jobID, "")
		entry := chains[rootJob]
		if entry == nil {
			entry = &chainEntry{ids: map[string]bool{}}
			chains[rootJob] = entry
		}
		entry.ids[jobID] = true
		if jobID == rootJob {
			if closed, _ := record["chainClosed"].(bool); closed {
				entry.closed = true
			}
		}
		rank := 1
		if m := roundRank.FindStringSubmatch(jobID); m != nil {
			rank, _ = strconv.Atoi(m[1])
		}
		if !entry.hasNewest || rank > entry.newestRank {
			mtime := time.Time{}
			if info, err := os.Stat(path); err == nil {
				mtime = info.ModTime()
			}
			entry.hasNewest = true
			entry.newestRank = rank
			entry.status = status
			entry.mtime = mtime
		}
	}

	// An open chain is work in flight for a plan's purposes even between
	// rounds, but bounded by a grace window so an abandoned chain still goes
	// stale. jobsInFlight keeps the strict view; this is only for stale-plan
	// accuracy.
	current := map[string]bool{}
	for job := range live {
		current[job] = true
		current[roundSuffix.ReplaceAllString(job, "")] = true
	}
	grace := graceSeconds()
	nowTime := now()
	for rootJob, entry := range chains {
		if entry.closed || !entry.hasNewest {
			continue
		}
		if entry.status == "completed" && nowTime.Sub(entry.mtime).Seconds() <= grace {
			current[rootJob] = true
			for id := range entry.ids {
				current[id] = true
			}
		}
	}

	var lines []string
	for _, plan := range planFiles(root) {
		text, err := os.ReadFile(plan)
		if err != nil {
			continue
		}
		claim, ok := planField(string(text), "In flight right now")
		if !ok || claim == "" || unblockedField.MatchString(claim) {
			continue
		}
		name := relName(root, plan)
		if strings.Contains(strings.ToLower(claim), "gate") && gatesRunning(root) {
			continue
		}
		if len(current) == 0 {
			lines = append(lines, fmt.Sprintf("STALE-PLAN %s: claims work in flight while no job is running", name))
			continue
		}
		named := false
		for job := range current {
			if job != "" && strings.Contains(claim, job) {
				named = true
				break
			}
		}
		if !named {
			lines = append(lines, fmt.Sprintf("STALE-PLAN %s: names in-flight work that is not among the running jobs or open chains", name))
		}
	}
	return lines
}

func openWork(root string) []string {
	if jobsInFlight(root) > 0 {
		return nil
	}
	var lines []string
	for _, plan := range planFiles(root) {
		text, err := os.ReadFile(plan)
		if err != nil {
			continue
		}
		step, ok := planField(string(text), "Next step")
		if !ok || step == "" || settledStep.MatchString(step) {
			continue
		}
		if templateValue.MatchString(step) {
			lines = append(lines, fmt.Sprintf("TEMPLATE-UNFILLED %s: %s", relName(root, plan), step))
			continue
		}
		if waiting, ok := planField(string(text), "Waiting on the human"); ok && waiting != "" && !unblockedField.MatchString(waiting) {
			continue
		}
		lines = append(lines, fmt.Sprintf("OPEN-WORK %s: %s", relName(root, plan), step))
	}
	return lines
}

// planFiles lists the stream plans, sorted, excluding README.md (the format doc).
func planFiles(root string) []string {
	paths, _ := filepath.Glob(filepath.Join(root, "plans", "*.md"))
	sort.Strings(paths)
	var out []string
	for _, path := range paths {
		// goals.md is the goal ledger, read only by the goal parser
		// (scanner disjointness); README.md is the format doc.
		if base := filepath.Base(path); base != "README.md" && base != "goals.md" {
			out = append(out, path)
		}
	}
	return out
}

func relName(root, path string) string {
	if rel, err := filepath.Rel(root, path); err == nil {
		return rel
	}
	return path
}
